File: daily/2024/1652_Defuse_the_Bomb.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> decrypt(vector<int>& code, int k) {
        int n = code.size();
        vector<int> ans(n, 0);
        if (k == 0) return ans;
        bool negative = (k < 0);
        if (k < 0) k = -k;
        int acc = 0;
        for (int i = 0; i < k; i++) acc += code[i];
        for (int i = 0; i < n; i++){
            if (negative) ans[(i+k)%n] = acc;
            else ans[(i-1+n)%n] = acc;
            acc = acc - code[i] + code[(i+k)%n];
        }
        return ans;
    }
};

int main(){
    vector<int> code = {5,7,1,4};
    int k = 3;
    // vector<int> code = {1,2,3,4};
    // int k = 0;
    // vector<int> code = {2,4,9,3};
    // int k = -2;
    Solution S;
    vector<int> ans = S.decrypt(code, k);
    for (int i : ans) cout << i << " ";
    cout << endl;
    return 0;
}

File: daily_python/1652_Defuse_the_Bomb.py
import os
import sys
from typing import List

class Solution:
    def decrypt(self, code: List[int], k: int) -> List[int]:
        n = len(code)
        ans = [0] * n
        if k == 0:
            return ans
        acc = 0
        for i in range(abs(k)):
            acc += code[i]
        for i in range(n):
            acc -= code[i]
            acc += code[(i+abs(k))%n]
            if k > 0:
                ans[i] = acc
            else:
                ans[(i+abs(k)+1)%n] = acc
        return ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            code = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            k = int(lines[1].split('=')[1][:-1])
            dataset.append((code, k))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for code, k in dataset:
        results.append(solution.decrypt(code, k))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')